pkg/api/jobartifacts/query.go

```diff
@@ -159,5 +211,88 @@ func (q *JobArtifactQuery) getFileContentMatches(jobRunID int64, file *storage.O
 // ContentMatcher is a generic interface for matching content in artifact files
 type ContentMatcher interface {
 	// GetMatches reads lines from the provided reader and returns a content match object (possibly incomplete with an error)
-	GetMatches(reader *bufio.Reader) (interface{}, error)
+	GetMatches(reader *bufio.Reader) (MatchedContent, error)
+	// GetCacheKey uniquely identifies the content query so we can cache separate results of different queries against the same run
+	GetCacheKey() string
+	// GetContentTemplate returns an instance of the type returned in GetMatches (used to enable deserializing interface{} from cache)
+	GetContentTemplate() interface{}
+	// PostProcessMatch is called to trim (if needed) the full set of matches (which we cache) down to what was actually requested
+	PostProcessMatch(fullArtifact JobRunArtifact) JobRunArtifact
+}
+
+/*
+ Caching is implemented at the level of the job run, scoped to a specific file and content query.
+ Content match and file match results have an indicator for whether they timed out, and job runs have an indicator for
+ whether they are "final" meaning we have retrieved all the contents requested (possibly with errors, but not timeouts).
+
+ When a JAQ comes in for a job run, we check the cache for an existing entry for the job run ID + JAQ. If there is one
+ and it's complete, we can just return it. If an incomplete one exists, we can request just the parts that timed out,
+ and then cache that result. In this way, by re-querying the same job run, we can fill out the results until they are
+ completely cached.
+
+ For content matching, we will scan a file for the widest possible match, and then post-filter the results down to
+ what was requested. For example, if a string line matcher looks for "foo" with 2 lines of context and a limit of 2
+ matches, we will scan the file and cache 12 matches with 12 lines of context; then just return the 2 matches with
+ 2 lines of context. This way, adjusting the matcher limits will not require re-scanning the file or caching results at
+ different sizes.
+*/
+
+func (q *JobArtifactQuery) CacheKeyForJobRun(jobRunID int64) string {
+	key := map[string]string{
+		"type":     "JAQJobRun~v1",
+		"id":       strconv.FormatInt(jobRunID, 10),
+		"pathGlob": q.PathGlob,
+	}
+	if q.ContentMatcher != nil {
+		key["contentMatcher"] = q.ContentMatcher.GetCacheKey()
+	}
+
+	jsonBytes, err := json.Marshal(key)
+	if err != nil {
+		log.Errorf("CacheKeyForJobRun should never fail to serialize the key: %v", err)
+		panic(err) // an engineer causing this somehow should find out ASAP
+	}
+	return string(jsonBytes)
+}
+
+const cacheExpiration = 4 * time.Hour
+
+func (q *JobArtifactQuery) SetJobRunCache(ctx context.Context, jobRunID int64, response JobRun) error {
```

Comment:
Caching is just the match data, not full files that matched, do I have that right?

Reply:
correct.